wal/src/lib.rs

```diff
@@ -241,113 +259,159 @@ impl Wal {
+    /// Gets a watch channel that will notify when the buffered ops are flushed to disk
+    pub fn notify_on_flush(&self) -> watch::Receiver<Option<WriteResult>> {
+        let b = self.buffer.lock().expect("mutex poisoned");
+        b.flush_notification.clone()
     }
-}
 
-/// Handle to rotate open segments to closed and delete closed segments.
-#[derive(Debug)]
-pub struct WalRotator<'a>(&'a Wal);
-
-impl<'a> WalRotator<'a> {
     /// Closes the currently open segment and opens a new one, returning the closed segment details.
-    pub async fn rotate(&self) -> Result<ClosedSegment> {
-        let closed = OpenSegmentFile::one_command(
-            &self.0.open_segment.tx.clone(),
-            OpenSegmentFileWriterRequest::Rotate,
-            (),
-        )
-        .await?;
-        let previous_value = self
-            .0
-            .closed_segments
-            .write()
-            .await
-            .insert(closed.id, closed.clone());
+    pub fn rotate(&self) -> Result<ClosedSegment> {
+        let new_open_segment =
+            OpenSegmentFileWriter::new_in_directory(&self.root, Arc::clone(&self.next_id_source))
+                .context(UnableToCreateSegmentFileSnafu)?;
+
+        let mut segments = self.segments.lock().expect("mutex poisoned");
+
+        let closed = std::mem::replace(&mut segments.open_segment, new_open_segment);
+        let closed = closed.close().expect("should convert to closed segmet");
+
+        let previous_value = segments.closed_segments.insert(closed.id(), closed.clone());
         assert!(
             previous_value.is_none(),
             "Should always add new closed segment entries, not replace"
         );
+
         Ok(closed)
     }
 
+    async fn flush_buffer_background_task(&self) {
+        let mut interval = tokio::time::interval(WAL_FLUSH_INTERVAL);
+
+        loop {
+            interval.tick().await;
+
+            // only write to the disk if there are writes buffered
+            let filled_buffer = {
+                let mut b = self.buffer.lock().expect("mutex poisoned");
+                if b.ops.is_empty() {
+                    continue;
+                }
+
+                std::mem::replace(&mut *b, WalBuffer::new())
+            };
+
+            // do the encoding while we're not holding any locks
+            let ops: Vec<_> = filled_buffer
+                .ops
+                .into_iter()
+                .map(ProtoSequencedWalOp::from)
+                .collect();
+            let batch = ProtoWalOpBatch { ops };
+            let encoded = batch.encode_to_vec();
+
+            // now write the data and let everyone know how things went
+            let mut segments = self.segments.lock().expect("mutex poisoned");
+            let res = match segments.open_segment.write(&encoded) {
+                Ok(summary) => WriteResult::Ok(summary),
+                Err(e) => WriteResult::Err(e.to_string()),
+            };
```

Comment:
Calling `send()` on `notify_flush` involves acquiring another mutex which might block. We can avoid holding the `segments` lock while we do that:

```suggestion
            let res = {
                let mut segments = self.segments.lock().expect("mutex poisoned");
                match segments.open_segment.write(&encoded) {
                    Ok(summary) => WriteResult::Ok(summary),
                    Err(e) => WriteResult::Err(e.to_string()),
                }
            };
```